### PR TITLE
Dynamically retrieve Linux runtimes

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/Microsoft/vscode-azureappservice/issues/724

I thought leaving the static list might be worthwhile to keep in case the request or JSON parse fails.  Thoughts?

### CODE REVIEW

File: appservice/src/createAppService/SiteRuntimeStep.ts [hunk 1]
```diff
@@ -3,18 +3,35 @@
...
+import * as request from 'request-promise';
 import { workspace } from 'vscode';
 import { AzureWizardPromptStep, IAzureQuickPickItem } from 'vscode-azureextensionui';
 import { ext } from '../extensionVariables';
 import { localize } from '../localize';
+import { signRequest } from '../signRequest';
 import { AppKind, WebsiteOS } from './AppKind';
 import { IAppServiceWizardContext } from './IAppServiceWizardContext';
 
 interface ILinuxRuntimeStack {
```
Comment (EricJizbaMSFT) [inline appservice/src/createAppService/SiteRuntimeStep.ts]: I don't like having `ILinuxRuntimeStack` _and_ the type in `availableStacksJson`. Can we consolidate? Would prefer the same type as the portal

File: appservice/src/createAppService/SiteRuntimeStep.ts [hunk 2]
```diff
@@ -55,8 +72,41 @@ export class SiteRuntimeStep extends AzureWizardPromptStep<IAppServiceWizardCont
...
+            return this.getLinuxRuntimeStackBackup();
+        }
+    }
+
+    private sortQuickPicksByRuntime(runtimeItems: IAzureQuickPickItem<ILinuxRuntimeStack>[], recommendedRuntimes: string[]): IAzureQuickPickItem<ILinuxRuntimeStack>[] {
+        function getPriority(item: IAzureQuickPickItem<ILinuxRuntimeStack>): number {
+            const index: number = recommendedRuntimes.findIndex((runtime: string) => item.data.name.includes(runtime));
+            return index === -1 ? recommendedRuntimes.length : index;
+        }
+        return runtimeItems.sort((a: IAzureQuickPickItem<ILinuxRuntimeStack>, b: IAzureQuickPickItem<ILinuxRuntimeStack>) => getPriority(a) - getPriority(b));
```
[the remark this answers is not in the document]
Comment (nturinski) [reply appservice/src/createAppService/SiteRuntimeStep.ts]: It doesn't.  The response is pretty similar to how we had stactically defined it:

image

File: appservice/src/createAppService/SiteRuntimeStep.ts [hunk 3]
```diff
@@ -55,8 +72,41 @@ export class SiteRuntimeStep extends AzureWizardPromptStep<IAppServiceWizardCont
...
+        requestOptions.headers = {
+            ['User-Agent']: 'vscode-azureappservice-extension'
+        };
+        requestOptions.url = 'https://management.azure.com/providers/Microsoft.Web/availableStacks?osTypeSelected=Linux&api-version=2018-02-01';
+        await signRequest(requestOptions, wizardContext.credentials);
+
+        try {
+            // returns a JSON-parseable string
+            // tslint:disable-next-line no-unsafe-any
+            const runtimes: string = await request(requestOptions).promise();
```
[the remark this answers is not in the document]
Comment (nturinski) [reply appservice/src/createAppService/SiteRuntimeStep.ts]: The OS is being passed in the request url. 

https://management.azure.com/providers/Microsoft.Web/availableStacks?_<b>osTypeSelected=Linux</b>_&api-version=2018-02-01

Additionally, _**Windows, WindowsFunctions,**_ and **_LinuxFunctions_** can all be passed in for `osTypeSelected`.

File: appservice/src/createAppService/SiteRuntimeStep.ts [hunk 4]
```diff
@@ -35,17 +40,18 @@ export class SiteRuntimeStep extends AzureWizardPromptStep<IAppServiceWizardCont
...
             wizardContext.newSiteRuntime = (await ext.ui.showQuickPick(runtimeItems, { placeHolder: 'Select a runtime for your new app.' })).data;
         } else if (wizardContext.newSiteOS === WebsiteOS.linux) {
-            let runtimeItems: IAzureQuickPickItem<ILinuxRuntimeStack>[] = this.getLinuxRuntimeStack().map((rt: ILinuxRuntimeStack) => {
+            /* tslint:disable:no-non-null-assertion */
+            let runtimeItems: IAzureQuickPickItem<ApplicationStack>[] = (await this.getLinuxRuntimeStack(wizardContext)).map((rt: ApplicationStack) => {
                 return {
-                    id: rt.name,
-                    label: rt.displayName,
+                    id: rt.name!,
+                    label: rt.display!,
```
Comment (EricJizbaMSFT) [inline appservice/src/createAppService/SiteRuntimeStep.ts]: nit: Would prefer `nonNullProp` over disabling `no-non-null-assertion`

**Pull request merged**